be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/ElectionValidator.scala

```diff
@@ -470,4 +535,131 @@ sealed class ElectionValidator(dbActorRef: => AskableActorRef) extends MessageDa
+  }
+
+  /** checks that the question ids received with the ResultElection are coherent with the ones received in the SetupElection
+    * @param rpcMessage
+    *   the rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param electionId
+    *   The electionId to compute the election question ids.
+    * @param error
+    *   The error to forward in case where ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultQuestionIds(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      electionId: Hash,
+      error: PipelineError
+  ): GraphMessage = {
+    val setupElectionIds = questions.map(_.id).toSet
+    val resultElectionIds = result.map(_.id).toSet
+    if (setupElectionIds == resultElectionIds) {
+      Right(rpcMessage)
+    } else
+      Left(error)
+  }
+
+  /** checks that for the given ElectionResult, the ballot options match the ballot options sent in the SetupElection message.
+    * @param rpcMessage
+    *   The rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param error
+    *   The error to forward in case where ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultBallotOptions(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      error: PipelineError
+  ): GraphMessage = {
+    var isResultBallotValid = true
+    result.foreach(electionQuestionResult => {
+      val matchingQuestion = findMatchingElectionQuestion(electionQuestionResult.id, questions)
+      if ((!matchingQuestion.isDefined) || (electionQuestionResult.result.toSet[ElectionBallotVotes].map(_.ballot_option) != matchingQuestion.get.ballot_options.toSet)) {
+        isResultBallotValid = false
+      }
+    })
+    if (isResultBallotValid) {
+      Right(rpcMessage)
+    } else
+      Left(error)
+  }
+
+  /** Finds the electionQuestion that matchs the given id.
+    * @param id
+    * @param questions
+    *   The list of election questions that has been sent with the setup Election message.
+    * @return
+    *   Some of the matching electionQuestion, or None if there is no ElectionQuestion that matches the given id.
+    */
+  private def findMatchingElectionQuestion(id: Hash, questions: List[ElectionQuestion]): Option[ElectionQuestion] = {
```

Comment:
You could remove this function if you use the `.find(_.id != id)` method. It returns the first element in the collection that satisfies the given predicate (supposing that all questions have different ids). It will either return an Option containing the matching element or None if no element is found.

Reply:
Now you can remove this function, since you do not use it anywhere anymore.